src/AccessLogEntry.cc

```diff
@@ -88,3 +88,12 @@ AccessLogEntry::~AccessLogEntry()
 #endif
 }
 
+const SBuf *
+AccessLogEntry::effectiveVirginUrl() const
+{
+    const SBuf *effectiveUrl = request ? &request->url.absolute() : &virginUrlForMissingRequest_;
+    if (!effectiveUrl || effectiveUrl->isEmpty())
+        return nullptr;
```

Comment:
We should document why are we not consulting ALE::url here when the request is missing. I am not sure what the right explanation is, but can suggest the following:

```
    // ALE::url should not be set here either, but if it is set, then we missed
    // a setVirginUrlForMissingRequest() call somewhere, and that ALE::url may
    // not have been cleaned up according to our caller standards.
    return nullptr;
```